pkg/server/api/bundle/v1/service.go

```diff
@@ -296,7 +296,75 @@ func (s *Service) createFederatedBundle(ctx context.Context, b *types.Bundle, ou
 }
 
 func (s *Service) BatchUpdateFederatedBundle(ctx context.Context, req *bundle.BatchUpdateFederatedBundleRequest) (*bundle.BatchUpdateFederatedBundleResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method BatchUpdateFederatedBundle not implemented")
+	var results []*bundle.BatchUpdateFederatedBundleResponse_Result
+	for _, b := range req.Bundle {
+		results = append(results, s.updateFederatedBundle(ctx, b, req.InputMask, req.OutputMask))
+	}
+
+	return &bundle.BatchUpdateFederatedBundleResponse{
+		Results: results,
+	}, nil
+}
+
+func (s *Service) updateFederatedBundle(ctx context.Context, b *types.Bundle, inputMask, outputMask *types.BundleMask) *bundle.BatchUpdateFederatedBundleResponse_Result {
+	log := rpccontext.Logger(ctx).WithField(telemetry.TrustDomainID, b.TrustDomain)
+
+	td, err := spiffeid.TrustDomainFromString(b.TrustDomain)
+	if err != nil {
+		log.WithError(err).Error("Invalid request: trust domain argument is not valid")
+		return &bundle.BatchUpdateFederatedBundleResponse_Result{
+			Status: api.CreateStatus(codes.InvalidArgument, "trust domain argument is not valid: %q", b.TrustDomain),
+		}
+	}
+
+	if s.td.Compare(td) == 0 {
+		log.Error("Invalid request: updating a federated bundle for the server's own trust domain is not allowed")
+		return &bundle.BatchUpdateFederatedBundleResponse_Result{
+			Status: api.CreateStatus(codes.InvalidArgument, "updating a federated bundle for the server's own trust domain (%s) s not allowed", td.String()),
+		}
+	}
+
+	dsBundle, err := api.ProtoToBundle(b)
+	if err != nil {
+		log.WithError(err).Error("Failed to convert bundle")
+		return &bundle.BatchUpdateFederatedBundleResponse_Result{
+			Status: api.CreateStatus(codes.Internal, "failed to convert bundle: %v", err),
```

Comment:
is it really internal? it looks like an error on request

Reply:
Good point! we have been using `Internal` for this error in other RPCs. But now that you mention it, `InvalidArgument` seems like a better fit. I'll change it in the other RPCs as well for consistency.

Reply:
It should be `Internal` when converting from the `common` types back to a `types` type but `InvalidRequest` when converting from a `types` type to a `common` type. The former is internal because this means we have a bug in our conversion when there is no reason we shouldn't be able to convert. The latter is InvalidRequest because it is likely caused by bad API input.